File: src_py/passes/to_ssa.py
import json
import sys
from utils import cfg

class BasicBlock:
    def __init__(self, label: str = "", instrs: list = []):
        self.preds: list[BasicBlock] = []
        self.label: str = label
        self.instrs: list = instrs

class Phi:
    def __init__(self, block: BasicBlock):
        self.block: BasicBlock = block
        self.labels: list[str] = []
        self.arguments: list[str] = []
        pass

    def append_operand(self, label: str, rhs: str):
        self.labels.append(label)
        self.arguments.append(rhs)

cnt: int = 0
current_def: dict[str, str] = {}
incomplete_phis: dict[BasicBlock, dict[str, Phi]] = {}
sealed_blocks: set[BasicBlock] = set()
label_to_block = dict[str, BasicBlock] = {}

def write_variable(var: str, block: BasicBlock, val):
    current_def[var][block] = val

def read_variable(var: str, block: BasicBlock):
    if block in current_def[var]:
        # lvn case
        return current_def[var][block]
    # not locally defined, global value numbering
    return read_variable_recursive(var, block)

def read_variable_recursive(var: str, block: BasicBlock):
    if block not in sealed_blocks:
        val = Phi(block)
        incomplete_phis[block][var] = val
    elif len(block.preds) == 1:
        val = read_variable(var, block)
    else:
        val = Phi(block)
        write_variable(var, block, val)
        val = add_phi_operands(var, val)

def add_phi_operands(var: str, phi: Phi):
    for pred in phi.block.preds:
        phi.append_operand(pred.label, read_variable(var, pred))
    return try_remove_trivial_phi(phi)

def try_remove_trivial_phi(phi: Phi) -> Phi:
    # for now
    return phi

def seal_block(block: BasicBlock):
    for var, phi in incomplete_phis[block]:
        add_phi_operands(var, phi)
    sealed_blocks.add(block)


def should_keep(instr):
    # not all instructions have an 'op' field, like labels
    if "op" not in instr:
        return True
    return instr["op"] != "nop"

def basic_blocks_ssa(func):
    blocks: list[BasicBlock] = []
    curr_block: BasicBlock = None



    for inst in func["instrs"]:
        if "label" in inst:
            if curr_block:
                blocks.append(curr_block)
                curr_block = label_to_block.get(inst["label"], BasicBlock())
            curr_block.label = inst["label"]
            curr_block.instrs.append(inst)
        else:  # if not label, must be op
            if curr_block is None:  # no entry label
                curr_block = BasicBlock("", []) # empty label, add placeholders at end if needed
            curr_block.instrs.append(inst)
            if inst["op"] in cfg.branch_instructions:
                
                yield curr_block
                curr_block = None

    if curr_block:
        blocks.append(curr_block)
    return blocks

if __name__ == "__main__":
    prog = json.load(sys.stdin)
    for fn in prog["functions"]:
        fn["instrs"] = [instr for instr in fn["instrs"] if should_keep(instr)]
    json.dump(prog, sys.stdout, indent=2)

File: src/bin/to_ssa.rs
use std::collections::{HashMap, HashSet};
use std::hash::Hash;

use bril::bril_syntax::{BrilType, Instruction, Label, InstructionOrLabel, Program};
use bril::cfg::{BasicBlock, ConditionalTransferResult, CFG};
use serde_json::Value;

#[derive(Debug)]

pub struct VNTable<T> {
    // Since Rust is strict about matching types and doesn't allow a custom Phi object to inherit 
    // from Instruction, use numbers to represent instructions until substitution time
    // current_def and num_to_val chained together roughly approximates current_def in the paper

    var_to_num: HashMap<String, HashMap<usize, usize>>, // Variable to block_id, to value number in that block
    num_to_val: Vec<Instruction>,   // nonlocal value number to instruction. Note that `dest` is unused at this point
    sealed_blocks: HashSet<usize>,  // map of block_ids where block is sealed
    phi_to_block: HashMap<Instruction, usize>, // map of phi instructions to id of containing block
    incomplete_phis: HashMap<usize, HashMap<String, Instruction>>, // block_id to 
}

impl<T> VNTable<T> {
    // Local Value Numbering
    pub fn write_variable(&mut self, var: String, block: BasicBlock<T>, val: Instruction) {
        // TODO: slight inefficiency as clobbered variables aren't removed from the table 
        // all in all, it might be safe to assume num clobbered variables < overhead from hashmap
        // we could run experiments to verify
        // Also, we don't want to reuse the value number because doing so would violate 
        // the single assignment property
        let var_blocks = self.current_def.entry(var.clone()).or_insert_with(HashMap::new);
        var_blocks.insert(block, self.num_to_val.len());
        self.num_to_val.push(val);
    }

    // Rather than return the instruction, return the value number. 
    // Required for making phis work in Bril's syntax. 
    pub fn read_variable(&self, var: String, block: BasicBlock<T>) -> usize {
        if self.current_def.contains_key(&var) {
            // Defined within block; LVN scenario
            return Some(Some(self.current_def.get(&var)).expect("Impossible case reached!").get(&block)).expect("Impossible case reached!");
        }
        // Look for definition in predecessors
        return self.read_variable_recursive(&var, &block);
    }

    // Global Value Numbering
    pub fn read_variable_recursive(&mut self, var: String, block: BasicBlock<T>) -> usize {
        let mut val; 
        if !self.sealed_blocks.contains(&block) {
            // Incomplete CFG 
            val = self.new_phi(&block);
            self.incomplete_phis.get(&block).insert(var.clone(), &val);
        } else if block.predecessors.len() == 1 {
            // Trivial phi; can optimize
            let val_num = self.read_variable(var, block.predecessors.first());
        } else {
            val = self.new_phi(&block);
            self.write_variable(var, block, val);
            self.add_phi_operands(var, val);
        }
        self.write_variable(var, block, val);
        return val;
    }

    pub fn new_phi(&mut self, block: &BasicBlock<T>) -> Instruction {
        let phi = Instruction {
            op: "phi".to_string(),
            dest: None,
            args: Some(Vec::new()),
            labels: Some(Vec::new()),
            bril_type: None,
            value: None,
            funcs: None,
        };
        self.phi_to_block.insert(&phi, &block);
        return phi;
    }

    pub fn append_phi_operand(self, phi: Instruction, label: String, val: String) {
        assert!(&phi.is_phi());
        phi.args.unwrap_or_else(|| Vec::new()).push(val);
        phi.labels.unwrap_or_else(|| Vec::new()).push(label);
        // let mut args = phi.args.unwrap_or_else(|| Vec::new());
        // args.push(val);
        // let mut labels = phi.labels.unwrap_or_else(|| Vec::new());
        // labels.push(label);
    }

    //TODO: 
    pub fn add_phi_operands(&mut self, var: String, phi: Instruction) -> Instruction {
        for pred in self.phi_to_block.get(&phi) {
            let binding = self.read_variable_recursive(var, pred);
            self.append_phi_operand(phi, pred, binding);
        }
        return self.try_trivial_remove_phi(phi);
    }

    pub fn try_trivial_remove_phi(self, phi: Instruction) -> Instruction {
        // for now, try to get everything else to work
        return phi;
    }

    pub fn seal_block(&mut self, block: BasicBlock<T>) {
        for var in self.incomplete_phis.get(&block) {
            self.add_phi_operands(var, self.incomplete_phis.get(&block).get(&var));
        }
        self.sealed_blocks.insert(block);
    }

    // TODO: remove redundant phis using SCCs
    

}




fn main() {
    let prog = Program::stdin();

    let cfg = CFG::from_program(prog);

    // cfg.dataflow_forward_optimistically(forward_meet, forward_transfer, transform);
    let prog = cfg.to_program();

    prog.stdout()
    // cfg.print_hm();
}
